src/hotspot/share/opto/callnode.hpp

```diff
@@ -503,15 +503,76 @@ class SafePointNode : public MultiNode {
 };
 
 //------------------------------SafePointScalarObjectNode----------------------
-// A SafePointScalarObjectNode represents the state of a scalarized object
-// at a safepoint.
+// A SafePointScalarObjectNode (aka spobj) represents the state of one or more
+// scalarized objects at a SafePointNode. There are two scenarios where a
+// 'spobj' is created: 1) A scalar replaced (SR) object is directly referenced
+// by a SafePoint; 2) A scalar replaced object is participating in an allocation
+// merge (Phi) and the Phi is referenced by a SafePoint. The schematics of how
+// 'spobj' is used in both scenarios are described below.
```

Comment:
I am not comfortable with reusing SafePointScalarObjectNode for 2) since it describes totally different information.
I think it should be separate Node which points to array of  SFSO  id (in addition to Phis) similar how we do now if SFSO is referenced in other SFSO's field.   SFSO could be created before the merge. Consider:
```
  Point p = new Point();
  Point q = foo();
  if (cond) {
      q = p;
  }
  trap(p, q);
```

Reply:
I had considered that but decided not to do it to prevent adding a new IR node. I'll give that a shot and update this thread with how it goes.

Reply:
It **will** complicate your DebugInfo code (packing/unpacking) information. But I think it is right thing to do to avoid duplicated re-allocations during deoptimization.

Reply:
Hi @vnkozlov, just FYI. I made the changes that you suggested.